pkg/processor/schemaregistry/avro/schema.go

```diff
@@ -0,0 +1,325 @@
+		}
+		return avro.NewArraySchema(items), nil
+	case reflect.Map:
+		if t == structuredDataType {
+			// special case - we treat StructuredData like a struct
+			var fields []*avro.Field
+			valType := t.Elem()
+			for _, keyValue := range v.MapKeys() {
+				fs, err := reflectInternal(append(path, keyValue.String()), v.MapIndex(keyValue), valType)
+				if err != nil {
+					return nil, err
+				}
+				field, err := avro.NewField(keyValue.String(), fs)
+				if err != nil {
+					return nil, err
+				}
+				fields = append(fields, field)
+			}
+			rs, err := avro.NewRecordSchema(strings.Join(path, "."), "", fields)
+			if err != nil {
+				return nil, err
+			}
+			return rs, nil
+		}
+		if t.Key().Kind() != reflect.String {
+			return nil, cerrors.Errorf("%s: maps with key type %v not supported, avro only supports strings as keys", strings.Join(path, "."), t.Key().Kind())
+		}
+		// try getting value type based on the map type
+		if t.Elem().Kind() != reflect.Interface {
+			vs, err := reflectInternal(append(path, "value"), reflect.Value{}, t.Elem())
+			if err != nil {
+				return nil, err
+			}
+			return avro.NewMapSchema(vs), nil
+		}
+
+		// this is map[string]any, loop through all values and extracting their
+		// types into a union schema, null is included by default
+		types := []avro.Schema{&avro.NullSchema{}}
+		typesSet := make(map[[32]byte]struct{})
+		var valValue reflect.Value
+		for _, kv := range v.MapKeys() {
+			valValue = v.MapIndex(kv)
+			vs, err := reflectInternal(append(path, "value"), valValue, t.Elem())
+			if err != nil {
+				return nil, err
+			}
+			if _, ok := typesSet[vs.Fingerprint()]; ok {
+				continue
+			}
+			types = combineSchemaTypes(types, func(schema avro.Schema) bool {
+				fp := vs.Fingerprint()
+				if _, ok := typesSet[fp]; ok {
+					return true
+				}
+				typesSet[fp] = struct{}{}
+				return false
+			}, vs)
+		}
+		if len(v.MapKeys()) == 0 {
+			// it's an empty map, add string to types to have a valid schema
+			types = append(types, avro.NewPrimitiveSchema(avro.String, nil))
+		}
+		vs, err := avro.NewUnionSchema(types)
+		if err != nil {
+			return nil, cerrors.Errorf("%s: %w", strings.Join(path, "."), err)
+		}
+		return avro.NewMapSchema(vs), nil
+	case reflect.Struct:
+		var fields []*avro.Field
+		for i := 0; i < t.NumField(); i++ {
+			sf := t.Field(i)
+			name, ok := getStructFieldJSONName(sf)
+			if !ok {
+				continue // skip this field
+			}
+			var vfi reflect.Value
+			if v.IsValid() {
+				vfi = v.Field(i)
+			}
+			fs, err := reflectInternal(append(path, name), vfi, t.Field(i).Type)
+			if err != nil {
+				return nil, err
+			}
+
+			field, err := avro.NewField(name, fs)
+			if err != nil {
+				return nil, err
+			}
+			fields = append(fields, field)
+		}
+		rs, err := avro.NewRecordSchema(strings.Join(path, "."), "", fields)
+		if err != nil {
+			return nil, err
+		}
+		return rs, nil
+	}
+	// Invalid, Uintptr, UnsafePointer, Uint64, Uint, Complex64, Complex128, Chan, Func
```

Comment:
I'm probably missing something obvious, but why are uint64 and unit not supported, but unit32 for example is? Also, what's the reason for `complex` not being supported?

Reply:
There's no equivalent type in Avro for these types, the Avro library doesn't support them either. We could add support by marshaling them as [`fixed`](https://stackoverflow.com/a/75242092) or simply `bytes`, but we wouldn't know the original type when unmarshaling, we'd get a `[]byte` on the other side. Maybe custom types could be used to forward the type info.

Given that it's not trivial to add the support and given that they are not _that_ common I skipped them and I plan to add a note to the limitations of the processor.

Reply:
Makes sense to me. If not already, it would be good to have a word or two about that in the code.

Reply:
I added a comment to `extractor.Extract` about this.